Preserve MCP `_meta` and `audience` annotations in `MCPToolset` results

anirbanbasu opened the pull request:
Closes #3323 (per-part `_meta`) and #4353 (audience filtering).

Surfaces MCP tool result metadata and `audience` annotations through the new [`MCPToolset`][pydantic_ai.mcp.MCPToolset] (introduced in #5325), without leaking either to the LLM.

## Foundations (`messages.py` + `_mcp_util.py`)

- `metadata: dict[str, Any] | None` field on `BinaryContent`, `FileUrl` (+ subclasses), `UploadedFile` — the typed home for application-only payloads. Distinct from `vendor_metadata` (model-provider hints sent to the API); docstrings call this out explicitly.
- `ToolReturnPart.model_response_str` / `model_response_object` / `content_items` unwrap `TextContent` to its `.content` so per-part `metadata` doesn't leak to the model.
- `pydantic_ai/_mcp_util.py` — shared helpers: `is_for_assistant`, `partition_content`, `merge_meta`, `clean_meta` (strips the FastMCP `wrap_result` marker), `map_tool_result_part` (preserves per-part `_meta` on the mapped value), `map_tool_result_user_content` (for the audience=user side).

## `MCPToolset` integration

`MCPToolset.direct_call_tool` now:

- Maps top-level `result._meta` → `ToolReturn.metadata` (FastMCP's internal `fastmcp` namespace filtered out).
- Maps per-block `_meta` → the corresponding `TextContent.metadata` / `BinaryContent.metadata`. Embedded text resources shallow-merge outer + inner `_meta` (inner wins on shared keys).
- Partitions content blocks by `annotations.audience`: `'assistant'`/unset parts become the `ToolReturn.return_value` visible to the model; `'user'`-only parts flow back as `ToolReturn.content` (typed `Sequence[UserContent]`) on a separate `UserPromptPart`.

When a result has no `_meta` and no user-audience parts, `direct_call_tool` returns the raw mapped value — same shape as before. `ToolReturn` is only synthesized when one of the new signals is present, so existing callers see no behavior change.

## Deprecated `MCPServer*` / `FastMCPToolset` — intentionally not backported

Both class hierarchies are `@deprecated` and their deprecation messages already point at `MCPToolset`. With v2 imminent, the maintenance cost (cassettes, two more code paths with subtle protocol differences) isn't worth the marginal v1 benefit — users wanting MCP metadata access can migrate to `MCPToolset` (drop-in replacement per the deprecation messages).

---

Originally opened by @anirbanbasu (per-part `_meta` extraction); audience-filtering work from the closed #4526 by @adityasingh2400 is folded in.

anirbanbasu commented:
@DouweM this is the PR in response to #3323 -- I am still working on it.

anirbanbasu commented:
Hi @DouweM, quoting you from https://github.com/pydantic/pydantic-ai/issues/3323#issuecomment-3482848716, I am pondering over whether it is better to edit the `_map_tool_result_part` in the `MCPServer` to handle the multi-modal content or modify the `_call_tool` method in the `_agent_graph.py`.

At present, if you take a look at the `_map_tool_result_part` method in this PR, I have only modified it to include the metadata if it is `TextContent`. I think I must also add the metadata if the response has multi-modal content. So, perhaps, changing this method to handle the multi-modal content is better instead of editing the `_call_tool` method.

anirbanbasu commented:
Quoting myself (https://github.com/pydantic/pydantic-ai/pull/3339#issuecomment-3490844770)
> [...] perhaps, changing this [i.e., `_map_tool_result_part `] method to handle the multi-modal content is better instead of editing the _call_tool method

Note that I just did that @DouweM. Need to add more tests to improve coverage.

DouweM commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -254,9 +254,20 @@ async def direct_call_tool(
-            return structured
+            if isinstance(structured, dict) and (
+                (len(structured) == 1 and 'result' in structured)
+                or (len(structured) == 2 and 'result' in structured and '_meta' in structured)
```
I don't think the metadata is supposed to be _on_ the `result.structuredContent`. Wouldn't it be on `result.meta`?

anirbanbasu replied:
I agree with you regarding not modifying the structured result.

We may be, though, not thinking of the same metadata.

What you are referring to is a FastMCP tool call result metadata. In addition, the `meta` is an attribute of FastMCP `ToolResult` only from version 2.13.1 (according to https://gofastmcp.com/servers/tools#toolresult-and-metadata) while the version currently in use with Pydantic AI is 2.12.4.

What I have been referring to is the `_meta` in the latest MCP standard https://modelcontextprotocol.io/specification/2025-06-18/basic/index#meta. FastMCP wraps this in `TextContent` and other types of content too.

If we go with the FastMCP-specific `meta` then there is a possibility that a MCP server implemented without using that specific version (> 2.13.1) of FastMCP or implemented in a different language will not return the metadata in the expected `ToolResult` style object.

Having said that, there is a possibility that FastMCP is implementing what the upcoming MCP standard will be, as they seem to typically do. (I haven't dug through this in details.)

In summary:
 1. for structured content, I think we could go with `meta` of `ToolResult` but I need to upgrade FastMCP for Pydantic AI to 2.13.1 or above;
 2. however, we ought to support `_meta` aliased `meta` for each content type.

Regarding (1), is this something I should do by myself?

What are your thoughts?

anirbanbasu replied:
Fixed this in commit 39d47b5363ea42792967ac615a079372860c1027.

anirbanbasu replied:
Please see simplified parsing in the latest commit.

anirbanbasu commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -374,35 +385,87 @@ async def _sampling_callback(
         # See https://github.com/jlowin/fastmcp/blob/main/docs/servers/tools.mdx#return-values
 
+        # Let's also check for metadata but it can be present in not just TextContent
+        metadata: dict[str, Any] | None = part.meta
```
I see your point in https://github.com/pydantic/pydantic-ai/issues/2288#issuecomment-3184826294 -- the same applies to tool calls.

However, maybe, we still should not ignore the `_meta` in content blocks because maybe the Pydantic AI user has some use of them (other than passing these to the LLM), such as logging.

💡 That makes me think that maybe there is a need to support _a user-specified MCP tool/feature call handler_, using which, the user could figure out a way to handle tool call in a way they want while Pydantic AI's default handler will ignore `_meta` but support FastMCP `ToolResult.meta`.

What do you think @DouweM?

anirbanbasu commented:
@DouweM thanks a lot for your thoughts. Here are my responses.

> 1. There should be a way to access both tool-result metadata and tool-result-item metadata

Yes, good idea. However, to pass the tool-result metadata, one needs to use FastMCP 2.13.1 (as mentioned in https://gofastmcp.com/servers/tools#toolresult-and-metadata) but this version is yet to be released. Thus, while I can parse tool-result metadata using the `CallToolResult`, I cannot create the tool to pass the metadata for `pytest` testing -- unless Pydantic AI uses the latest from the default branch of FastMCP's GitHub, which is a bad idea, in my understanding.

_Related to this, I am attaching tool-result metadata using FastMCP's latest from GitHub in my [MCP server template](https://github.com/anirbanbasu/pymcp/blob/master/src/pymcp/mixin.py#L64)._

> 2. MCP supports `_meta` on every object, so result.meta is not FastMCP specific, it's right in the MCP SDK: [`CallToolResult`](https://github.com/modelcontextprotocol/python-sdk/blob/9eae96a05e09fa3f35ea50f8b1c11b19c69f9597/src/mcp/types.py#L916) inherits from Result which has a [`meta` field](https://github.com/modelcontextprotocol/python-sdk/blob/9eae96a05e09fa3f35ea50f8b1c11b19c69f9597/src/mcp/types.py#L108) corresponding to `_meta`.

Yes, indeed. Thanks for pointing out. That link effectively points to [this documentation in a comment](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields), which is the same as the [June standard](https://modelcontextprotocol.io/specification/2025-06-18/basic/index#meta).

> (2) i. I think this is the natural place to read "tool-result metadata" from (and for any MCP server SDK to write metadata)

Okay, but is it (i.e., [`_meta`](https://modelcontextprotocol.io/specification/2025-06-18/basic/index#meta)) also referring to the tool-result-item metadata?

> (2) ii. I don't think we should ever try to extract metadata from result.structuredData, as we should treat it as arbitrary data

Yes, that was my mistake. Apologies. I will fix it. If the data is structured data then I can attach the metadata as I am doing in my [MCP server template](https://github.com/anirbanbasu/pymcp/blob/master/src/pymcp/mixin.py#L64).

> 3. If there is only tool-result metadata and no tool-result-item metadata, `MCPServer.direct_call_tool` should return it as `ToolReturn.metadata`

Okay, makes sense.

> 4. If there is no tool-result metadata, but there is a single tool-result-item (`TextContent` etc) that does have `.meta`, that should be the `ToolReturn.metadata`

Okay, that's reasonable.

> 5. If there are multiple pieces of metadata, on the tool-result and/or on one or more tool-result-items, we have a couple of options. Since `ToolReturn.content` will hold the tool-result-items in this case, perhaps `ToolReturn.metadata` should be a dictionary like `{'result': <result metadata>, 'content': [<metadata for content index 0>, ...]]}`
> i. Another option is to leverage the existing process_tool_call hook, but (unfortunately?) it gets the processed tool result rather than the raw thing. Perhaps it could pass a flag into `direct_call_tool` to return the raw `mcp.types.CallToolResult` so metadata can be read off of it?

The first option sounds reasonable as a default option even if the dictionary keys seem a bit hard-coded.

Instead of your suggested second option, how about letting the `process_tool_call` hook receive the processed `ToolResult` but alter it as necessary? This should be possible because the `ToolReturn.metadata` is just a dictionary after all.

anirbanbasu commented:
@DouweM with the latest commit (39d47b5363ea42792967ac615a079372860c1027), I have attempted to implement 3, 4 and 5 from your message above (https://github.com/pydantic/pydantic-ai/pull/3339#issuecomment-3503048730).

Notes.

 - I modified 5 a bit to return a dictionary instead of a list because there may be content parts that do not have corresponding metadata.
 - Modified `_map_tool_result_part` to return a tuple so that the second item in the tuple is the part metadata if it exists.
 - I am not convinced if the code (see snippet below, lines 303-309 of `mcp.py`) to handle the part content if it is not multi-modal is what is intended.
 ```python
if isinstance(mapped_part, messages.BinaryContent):
    identifier = mapped_part.identifier

    return_values.append(f'See file {identifier}')
    user_contents.append([f'This is file {identifier}:', mapped_part])
else:
    user_contents.append(mapped_part)
 ```
 - Tests will definitely fail because they have not been correctly written yet.
 - Coverage will also fail for the same reason.

anirbanbasu commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -374,35 +385,87 @@ async def _sampling_callback(
     async def _map_tool_result_part(
         self, part: mcp_types.ContentBlock
-    ) -> str | messages.BinaryContent | dict[str, Any] | list[Any]:
+    ) -> str | messages.ToolReturn | messages.BinaryContent | dict[str, Any] | list[Any]:
```
Rewrote the logic in commit 39d47b5363ea42792967ac615a079372860c1027.

anirbanbasu replied:
Please see simplified parsing in the latest commit.

DouweM commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -374,35 +442,32 @@ async def _sampling_callback(
-            return self._get_content(resource)
+            return self._get_content(part.resource), metadata
         elif isinstance(part, mcp_types.ResourceLink):
             resource_result: mcp_types.ReadResourceResult = await self._client.read_resource(part.uri)
```
Note that `resource_result` can also have its own `meta`, and so can each of its `contents` 😅 So we may need to build a single nested `metadata` dict from all of those

anirbanbasu replied:
I added a nested metadata parsing logic but: (a) I am not convinced that my assumption is entirely correct about `part.meta` being the same as `resource_result.meta` and (b) the code has metadata parsing logic, which could be deduplicated.

I will request your review.

DouweM commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -255,11 +255,79 @@ async def direct_call_tool(
 
-        mapped = [await self._map_tool_result_part(part) for part in result.content]
-        return mapped[0] if len(mapped) == 1 else mapped
+        mapped_part_metadata_tuple_list = [await self._map_tool_result_part(part) for part in result.content]
```
Let's call this `parts_with_metadata`

DouweM commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -255,11 +255,79 @@ async def direct_call_tool(
-        return mapped[0] if len(mapped) == 1 else mapped
+        mapped_part_metadata_tuple_list = [await self._map_tool_result_part(part) for part in result.content]
+        if (
+            all(mapped_part_metadata_tuple[1] is None for mapped_part_metadata_tuple in mapped_part_metadata_tuple_list)
```
I think this can be `all(metadata is None for (part, metadata) in mapped_part_metadata_tuple_list)`. Let's use tuple destructuring everywhere instead of index-based reading.

Also let's move this to a `parts_have_metadata = any(...)` variable (inverted to be positive) as we use it multiple times (e.g. `if not parts_have_metadata and result.metadata is None` reads pretty clearly)

DouweM replied:
@anirbanbasu Yep sounds good

DouweM commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -255,11 +255,79 @@ async def direct_call_tool(
+            # There is no metadata in the tool result or its parts, return just the mapped values
+            return (
+                mapped_part_metadata_tuple_list[0][0]
+                if len(mapped_part_metadata_tuple_list[0]) == 1
```
Shouldn't this be `len(mapped_part_metadata_tuple_list) == 1`?

DouweM replied:
I really don't like `[i]` for tuples, and we already know we have no part metadata here, so let's make this something like:

```py
parts = [part for (part, _) in parts_and_metadata]
# use parts as we did before
```

anirbanbasu replied:
I am extracting `parts_only` before getting into the `if-elif-else` blocks.

DouweM commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -255,11 +255,79 @@ async def direct_call_tool(
+                    # Merge the metadata dictionaries, with part metadata taking precedence
+                    if return_metadata.get('content', None) is None:
+                        # Create an empty list if it doesn't exist yet
+                        return_metadata['content'] = list[dict[str, Any]]()
```
We can likely use `return_metadata.setdefault('content', [])` just once before the `for`

anirbanbasu commented:
@DouweM updated the tests and included some `# pragma no-cover` directives to parts of the metadata processing code that can only be tested once FastMCP is upgraded to 2.13.1. I shall write those test once FastMCP is upgraded.

DouweM commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -273,12 +273,62 @@ async def direct_call_tool(
+
+        # The following branching cannot be tested until FastMCP is updated to version 2.13.1
+        # such that the MCP server can generate ToolResult and result.meta can be specified.
+        # TODO: Add tests for the following branching once FastMCP is updated.
```
Adding a comment here so we don't lose this

anirbanbasu replied:
I will add this in once we figure out about FastMCP 2.13.1 upgrade or if I can come up with a test whereby I construct the metadata by hand?

DouweM commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -273,12 +273,62 @@ async def direct_call_tool(
+        # The following branching cannot be tested until FastMCP is updated to version 2.13.1
+        # such that the MCP server can generate ToolResult and result.meta can be specified.
+        # TODO: Add tests for the following branching once FastMCP is updated.
+        if len(parts_metadata) > 0:
```
Here too, I prefer `if parts_metadata` over specifically length-checking, unless we want to treat things like None and empty differently

anirbanbasu replied:
Please see simplified parsing in the latest commit.

anirbanbasu commented:
@DouweM okay, I will modify the `BinaryContent` as discussed and keep parts metadata with the parts themselves.

However, we will run into an issue with `TextContent` because the current code discards its metadata and this cannot be packaged as `BinaryContent`. Perhaps, we should output `TextContent` as is?

> > Lastly, is there a plan to upgrade FastMCP to 2.13.1, now that it is an officially released version?

> Feel free to update!

It seems that updating it to 2.13.1 is impossible for now due to other dependencies.
```sh
% uv lock -P fastmcp --dry-run                   
Resolved 381 packages in 191ms
Update fastmcp v2.12.4 -> v2.12.5
```
This would mean that an exhaustive test set cannot be implemented until we can get 2.13.1.

anirbanbasu committed:
chore: Started off with capturing metadata for structured result. More to come!

anirbanbasu committed:
Merge branch 'main' into patch-3323

anirbanbasu committed:
feat: Added parsing of metadata, nested or not, in binary and text content including resources.
chore: Updated tests but have not added tests specific to metadata, hence coverage should fail.

anirbanbasu committed:
chore: Added top-level ToolReturn metadata.

anirbanbasu commented on pydantic_ai_slim/pydantic_ai/_agent_graph.py:
```diff
@@ -1110,6 +1110,7 @@ async def _call_tool(
                     f'The return value of tool {tool_call.tool_name!r} contains invalid nested `ToolReturn` objects. '
                     f'`ToolReturn` should be used directly.'
                 )
+            # TODO: Keep updated with the binary parsing in mcp.py
```
Do we need this anymore because I don't construct `BinaryContent` in that way anymore?

anirbanbasu commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -273,12 +273,62 @@ async def direct_call_tool(
+
+        parts_with_metadata = [await self._map_tool_result_part(part) for part in result.content]
+        parts_only = [part for part, _ in parts_with_metadata]
+        # any_part_has_metadata = any(metadata is not None for _, metadata in parts_with_metadata)
```
No, it's gone. Please see simplified parsing in the latest commit.

anirbanbasu commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -273,12 +273,62 @@ async def direct_call_tool(
+        # such that the MCP server can generate ToolResult and result.meta can be specified.
+        # TODO: Add tests for the following branching once FastMCP is updated.
+        if len(parts_metadata) > 0:
+            if result.meta is not None and len(result.meta) > 0:  # pragma: no cover
```
Okay, I have no no-cover but that's because I have not written the tests.

anirbanbasu commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -273,12 +273,62 @@ async def direct_call_tool(
+                else:
+                    return_metadata = {'content': parts_metadata}  # pragma: no cover
+        else:
+            if result.meta is not None and len(result.meta) > 0:  # pragma: no cover
```
Please see simplified parsing in the latest commit.

anirbanbasu commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -394,35 +444,57 @@ async def _sampling_callback(
-                if len(resource_result.contents) == 1
-                else [self._get_content(resource) for resource in resource_result.contents]
-            )
+            # Check if metadata already exists. If so, merge it with nested the resource metadata.
```
This doesn't exist anymore. Please see simplified parsing in the latest commit.

anirbanbasu commented on tests/test_mcp.py:
```diff
@@ -87,6 +88,39 @@ async def test_stdio_server(run_context: RunContext[int]):
         assert result == snapshot(32.0)
 
 
+async def test_tool_response_metadata(run_context: RunContext[int]):
```
I shall write new tests.

anirbanbasu committed:
test: Added first metadata test with a single text part.

anirbanbasu committed:
chore: Added a no-cover directive and a note why ToolResult.meta cannot be tested until FastMCP can be upgraded to 2.13.1.

DouweM commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -490,12 +490,23 @@ async def direct_call_tool(
+            return (  # pragma: no cover
+                messages.ToolReturn(return_value=mapped[0], metadata=result.meta)
+                if len(mapped) == 1
+                else messages.ToolReturn(return_value=mapped, metadata=result.meta)
```
This will cause us to run into trouble here, because we're putting `BinaryContent` inside `ToolReturn.return_value`:

https://github.com/pydantic/pydantic-ai/blob/c27e5e4e32f16d2e28c3b25a10528f64335e7cfa/pydantic_ai_slim/pydantic_ai/_agent_graph.py#L1161-L1172

So we should make that logic less strict, so that if it finds a `ToolReturn.return_value` with `BinaryContent` in it, it'll go through this same logic to move them to the `content` field:

https://github.com/pydantic/pydantic-ai/blob/c27e5e4e32f16d2e28c3b25a10528f64335e7cfa/pydantic_ai_slim/pydantic_ai/_agent_graph.py#L1140-L1159

(Although part of me is thinking that maybe I should do https://github.com/pydantic/pydantic-ai/issues/3253 first, because that's going to affect so much of this logic and likely make it easier, as we can keep binary parts on `ToolReturn.return_value` without having to move them... If you can wait with this PR, I could maybe look at that one first?)

We should also update the tests to not just test `direct_call_tool`, but actually test using these tools in an agent run, to ensure we go through that part of the agent graph tool call result parsing code.

anirbanbasu committed:
Merge branch 'main' into patch-3323

anirbanbasu commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -490,12 +490,23 @@ async def direct_call_tool(
+                else messages.ToolReturn(return_value=mapped, metadata=result.meta)
+            )
+        else:
+            return mapped[0] if len(mapped) == 1 else mapped
```
Done.

anirbanbasu commented on pydantic_ai_slim/pydantic_ai/mcp.py:
```diff
@@ -682,24 +695,29 @@ async def _sampling_callback(
-                    pass
-            return text
+            if part.meta:
+                return messages.TextPart(content=text, metadata=part.meta)
```
Okay, I see. Then should `messages.TextContent` and messages.TextPart` look exactly the same except that `messages.TextContent` can have `metadata` and it should not have `id` indicating an identifier for the part?

anirbanbasu committed:
feat: Added a TextContent.

anirbanbasu reviewed (commented):
Added `TextContent` but it required a lot of checks in the `pydantic_ai.models`.

anirbanbasu commented:
> Added `TextContent` but it required a lot of checks in the `pydantic_ai.models`.

This is a very messy change. I would rather roll it back if we could. Do you have any suggestions @DouweM ?

DouweM commented:
@anirbanbasu Yeah I hate to say it because we've going through a few versions of this already, but I think it'll be better to do https://github.com/pydantic/pydantic-ai/issues/3253 first (hopefully this month) and then come back to this. That'll also give me some time to think on the "metadata on text" case. I think subclassing `str` would be worth trying out, so we don't have to modify all the models in their handling of `UserPromptPart.content`.

anirbanbasu committed:
Merge branch 'main' into patch-3323

DouweM commented:
Sorry for the radio silence here, the PR will need a bunch of work so I think it's easiest for me to take it over from here!

anirbanbasu commented:
@DouweM thanks for the update, looking forward to integrating this MCP metadata extraction functionality!

DouweM committed:
Merge remote-tracking branch 'origin/main' into patch-3323

# Conflicts:
#	pydantic_ai_slim/pydantic_ai/mcp.py
#	pydantic_ai_slim/pydantic_ai/messages.py
#	pydantic_ai_slim/pydantic_ai/models/bedrock.py
#	pydantic_ai_slim/pydantic_ai/models/gemini.py
#	pydantic_ai_slim/pydantic_ai/models/google.py
#	pydantic_ai_slim/pydantic_ai/models/huggingface.py
#	pydantic_ai_slim/pydantic_ai/models/openai.py
#	tests/test_agent.py

DouweM committed:
Add programmatic `metadata` field to `BinaryContent`, `FileUrl` (and subclasses), and `UploadedFile`

Mirrors the existing `metadata` on `TextContent` for app-level data
that's not sent to the LLM. Distinct from `vendor_metadata`, which
carries provider-specific hints used in API requests.

Will be populated from MCP `_meta` on per-part content blocks in a
follow-up commit.

DouweM committed:
Surface MCP `_meta` and audience filtering on `MCPServer` and `FastMCPToolset`

Wraps tool results in `messages.ToolReturn` only when there's metadata
to attach:

- `ToolReturn.metadata['meta']` — top-level `CallToolResult._meta`.
- `ToolReturn.metadata['user_content']` — content blocks annotated
  `audience=['user']` (as `ContentBlock.model_dump()` dicts), so the
  application sees them but the model doesn't.

Per-content-block `_meta` rides on `TextContent.metadata` /
`BinaryContent.metadata` directly so callers can read it without
unpacking dicts. `ResourceLink` is eagerly dereferenced and its
`_meta` is merged onto the resolved content.

The `_mcp_util.py` helpers (audience partition, metadata builder,
meta merge) are shared between both MCP implementations.

`output_schema` on `ToolDefinition.metadata` is now redundant with the
top-level `return_schema` field; flagged for removal in v2.

Co-authored-by: Aditya Singh <adityasingh2400@users.noreply.github.com>

DouweM committed:
Tests: MCP `_meta` extraction and audience filtering

Adds fixture tools to `tests/mcp_server.py` (`get_with_result_meta`,
`get_user_only_audience`, `get_mixed_audience`, `get_image_with_meta`)
and matching tools in the in-memory `tests/test_fastmcp.py` server.

`tests/test_mcp.py` and `tests/test_fastmcp.py` cover:

- per-part `_meta` on text → `messages.TextContent.metadata`
- per-part `_meta` on image → `messages.BinaryImage.metadata`
- top-level `CallToolResult._meta` → `ToolReturn.metadata['meta']`
- audience-only-user content → `ToolReturn.metadata['user_content']`
  with placeholder `return_value`
- mixed audience → assistant text on `return_value`, user-only on
  `metadata['user_content']`

Co-authored-by: Aditya Singh <adityasingh2400@users.noreply.github.com>

DouweM committed:
Docs: MCP tool result metadata and audience filtering

DouweM committed:
Update output schema snapshots for new `BinaryContent.metadata` field

DouweM committed:
Address auto-review feedback: stop `TextContent.metadata` from reaching the LLM

Fixes the leak flagged by both `devin` and `Copilot`: when an MCP tool
returned a `TextContent` with per-part `_meta` and no other metadata
to wrap, the bare `TextContent` ended up in `ToolReturnPart.content`
and `tool_return_ta.dump_json` serialized the whole dataclass to the
model, including `metadata` and `kind`. `ToolReturnPart.model_response_str`
/ `model_response_object` / `content_items` now unwrap `TextContent`
to its `.content` string, matching the existing user-prompt convention
in every model adapter.

Also:

- Move the `_mcp_util` import in `toolsets/fastmcp.py` inside the
  guarded `try/except ImportError` so missing-dep environments still
  see the friendly install message.
- `_attach_meta` now returns a fresh `replace(item, metadata=...)`
  instead of mutating its argument.
- Inline the `USER_ONLY_PLACEHOLDER_TEXT` literal in tests rather than
  importing from a private module.
- Comment the FastMCP `fastmcp.wrap_result` sanitization so the
  `{} -> None` fallback is no longer mysterious.
- Docs: add a code example showing how to read `_meta` and
  `user_content` off the message history, and note that
  `read_resource` may return `TextContent` (not just `str`) when the
  underlying resource carries `_meta`.

DouweM committed:
Skip ruff on the doc example with top-level `await`

`F704 await statement outside of a function` was breaking
`test_docs_examples` for the new `mcp_tool_result_metadata.py` example.
Following the docs guideline of fence-level `{test="skip" lint="skip"}`
over inline suppressions, and wrapping the example body in an async
function for readability.

DouweM committed:
Cover the remaining metadata-flow paths

CI's coverage report flagged 10 lines untouched by the existing test
suite. Adds:

- `get_text_resource_with_meta`, `get_image_resource_link_with_meta`,
  and `get_resource_link_with_meta` fixture tools and their tests for
  the EmbeddedResource-with-`_meta`, ResourceLink-with-`_meta` →
  TextContent, and ResourceLink-with-`_meta` → BinaryContent paths.
- `embedded_text_resource_with_meta` fixture + test on the FastMCP
  side, covering its embedded-text-with-`_meta` branch.
- A regression test asserting `TextContent.metadata` does not leak
  through `ToolReturnPart.model_response_str` /
  `model_response_object`.
- A monkeypatched test for `_read_resource_link`'s list branch (when
  `read_resource` returns multiple contents).

DouweM commented:
*[Drafted by Claude on @DouweM's behalf — context is the v2 MCP overhaul we're scoping in another worktree.]*

Catching up here. Two distinct pieces in this PR that should be split:

**Part 1 — `messages.py` extensions + shared helpers (low-risk, mergeable):**
- New `metadata: dict[str, Any] | None` field on `BinaryContent`, `FileUrl` (+ subclasses), `UploadedFile`
- `ToolReturnPart.content_items()` / `model_response_str()` / `model_response_object()` stripping `TextContent.metadata` from the model-facing payload
- New `pydantic_ai/_mcp_util.py` with `partition_content` / `build_tool_return_metadata` / `merge_meta` / `is_for_assistant`

Forward-compatible regardless of which MCP toolset wraps the call. Worth splitting into a precursor PR and landing first.

**Part 2 — MCP integration plumbing in `pydantic_ai/mcp.py` and `pydantic_ai/toolsets/fastmcp.py`:**

This is going to need rework. Both of those classes are about to be `@deprecated` in the v2-prep PR and replaced by a new `pydantic_ai.mcp.MCPToolset`. Threading metadata through the legacy paths in detail is mostly wasted plumbing.

Before re-wiring, three design issues:

**1. `user_content` shape.** Currently `list[ContentBlock.model_dump()]` — raw MCP wire dicts in a public `metadata` payload. The whole point of `pydantic_ai.mcp.Resource` / `MCPError` / etc. was to *not* expose MCP SDK types. `user_content` should map through the same logic as `return_value` and end up as `list[str | TextContent | BinaryContent]` (or similar typed shape).

**2. `ToolReturn.metadata` is mixing two categories.** It carries both `meta` (free-form server `_meta`) and `user_content` (framework-synthesized, audience-filtered blocks). Apps have to remember which key came from where and whether it's typed. Cleaner: keep `metadata` as the server `_meta` passthrough only, and add a typed field on `ToolReturn` for user-only content — e.g. `ToolReturn.user_content: list[str | TextContent | BinaryContent] | None`. Then `metadata['user_content']` goes away.

**3. `BinaryContent.vendor_metadata` vs new `BinaryContent.metadata`.** Real distinction (model-vendor hints vs application opaque data) but not obvious. At minimum the docstring on the new field should call out the neighbor and explain the difference.

**Recommended sequencing:**
1. Split this PR — land the messages.py + `_mcp_util.py` precursor first.
2. Hold the MCP wiring until Phase A (new `MCPToolset`) lands.
3. Wire metadata exposure through the new `MCPToolset` with the typed `user_content` field on `ToolReturn`.
4. Decide whether to port a minimal version onto `MCPServer*` / `FastMCPToolset` for 1.x users on the deprecated paths, or skip given the v2 timeline.

DouweM committed:
Merge remote-tracking branch 'origin/main' into patch-3323

DouweM committed:
Narrow scope to Part 1 (messages.py + `_mcp_util.py` foundations)

Per the maintainer review (#3339#issuecomment-…): split this PR into a
forward-compatible precursor and defer the MCP integration plumbing
until the new `MCPToolset` (v2-prep Phase A) lands. The deferred work
also needs design changes: typed `user_content` field on `ToolReturn`
instead of a magic key in `metadata`, and `user_content` mapped through
the same logic as `return_value` rather than raw `ContentBlock.model_dump()`
dicts.

Kept (Part 1, mergeable):
- `metadata: dict[str, Any] | None` on `BinaryContent`, `FileUrl` (+
  subclasses), `UploadedFile`, with docstrings now calling out the
  distinction from neighbor `vendor_metadata`.
- `ToolReturnPart.{model_response_str,model_response_object,content_items}`
  unwrap `TextContent` to its `.content` so per-part `metadata` does
  not leak to the model.
- Shared MCP helpers in `pydantic_ai/_mcp_util.py` — `is_for_assistant`,
  `partition_content`, `merge_meta` — covered by direct unit tests in
  `tests/test_mcp_util.py` so they survive without callers.

Reverted (deferred to Part 2):
- All `pydantic_ai/mcp.py` and `pydantic_ai/toolsets/fastmcp.py`
  integration changes (audience filtering, top-level `_meta` wrapping,
  per-part meta plumbing, `_attach_meta`, `_read_resource_link` meta
  propagation).
- `build_tool_return_metadata` and `USER_ONLY_PLACEHOLDER_TEXT` from
  `_mcp_util.py` — both bake in the old design that's about to change.
- Fixture tools and integration tests that exercised the deferred
  plumbing.
- Docs section about MCP tool result metadata — defer until the wiring
  is in place.

DouweM committed:
Merge remote-tracking branch 'origin/main' into patch-3323

DouweM committed:
Part 2: surface MCP `_meta` + audience filtering through `MCPToolset`

Wires the Part 1 helpers (audience partition, `_meta` propagation) into the
new `MCPToolset` introduced by #5325:

- Top-level `result._meta` (FastMCP `fastmcp` marker filtered out) →
  `ToolReturn.metadata`
- Per-block `_meta` → `TextContent.metadata` / `BinaryContent.metadata`;
  embedded text resources shallow-merge outer + inner `_meta` (inner wins)
- `annotations.audience == ['user']` parts → `ToolReturn.content`
  (typed `Sequence[UserContent]`)

Results with no `_meta` and no user-audience parts return the raw mapped
value, same shape as before — `ToolReturn` is only synthesized when one of
the new signals is present.

The deprecated `MCPServer*` / `FastMCPToolset` classes intentionally aren't
backported: their deprecation messages already point at `MCPToolset`, and
the v2 timeline makes the maintenance cost (cassettes, two more code paths)
not worth it.

DouweM commented:
I'm making progress on this :) Can't say yet when it'll be released exactly.

anirbanbasu commented:
> I'm making progress on this :) Can't say yet when it'll be released exactly.

@DouweM is there anything I should do although you're handling it (https://github.com/pydantic/pydantic-ai/pull/3339#issuecomment-4300011317)?

DouweM commented:
@anirbanbasu Nope this is just blocked by me for the moment.

DouweM commented:
I'm about to go on parental leave (back on June 29), I'll continue on this when I'm back, appreciate the patience!

anirbanbasu commented:
@DouweM 

> I'm about to go on parental leave 

Congratulations on becoming a parent! Enjoy your leave.
